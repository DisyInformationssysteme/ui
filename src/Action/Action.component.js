import React, { PropTypes } from 'react';
import { Action as PureAction } from 'react-talend-components';
import { api } from 'react-cmf';
import invariant from 'invariant';

/**
 * @param {object} props react props
 * @example
<Action name="menu:demo"></Action>
 */
function Action(props, context) {
	const onClick = (event, payload) => {
		context.store.dispatch(payload.action);
	};
	let action;
	if (!props.action && props.name) {
		action = api.action.getActionInfo(context, props.name);
	} else if (props.action && !props.name) {
		action = props.action;
	} else {
		invariant(action, 'Action component should have name or action props');
	}
	return (
		<PureAction
			label={action.name}
			icon={action.icon}
			onClick={onClick}
			link={props.link}
			model={props.model}
			tooltipPlacement={props.tooltipPlacement}
			hideLabel={props.hideLabel}
		/>
	);
}

Action.propTypes = {
	name: PropTypes.string,
	action: PropTypes.shape({
		label: PropTypes.string,
		icon: PropTypes.string,
	}),
	link: PropTypes.boolean,
	model: PropTypes.object,
	tooltipPlacement: PropTypes.string,
	hideLabel: PropTypes.boolean,
};
Action.contextTypes = {
	router: PropTypes.object,
	store: PropTypes.object,
	registry: PropTypes.object,
};
export default Action;
