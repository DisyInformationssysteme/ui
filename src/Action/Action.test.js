import React from 'react';
import renderer from 'react-test-renderer';

import Action from './Action.component';

jest.mock('react-dom');

const myAction = {
	label: 'Click me',
	icon: 'fa fa-asterisk',
	onClick: jest.fn(),
};

describe('Action', () => {
	it('should render a button', () => {
		const wrapper = renderer.create(<Action {...myAction} />).toJSON();
		expect(wrapper).toMatchSnapshot();
	});
	it('should click on the button trigger the onclick props', () => {
		const wrapper = renderer.create(
			<Action extra="extra" {...myAction} />).toJSON();
		wrapper.props.onClick();
		expect(myAction.onClick).toHaveBeenCalled();
		expect(myAction.onClick.mock.calls.length).toBe(1);
		const args = myAction.onClick.mock.calls[0];
		expect(args.length).toBe(2);
		expect(args[0]).toBe();
		expect(args[1].action.extra).toBe('extra');
	});
	it('should pass all props to the Button', () => {
		const wrapper = renderer.create(
			<Action
				className="navbar-btn"
				notExisting
				{...myAction}
			/>).toJSON();
		expect(wrapper).toMatchSnapshot();
	});
});
