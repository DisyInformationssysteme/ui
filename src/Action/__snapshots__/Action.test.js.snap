exports[`Action should render its name 1`] = `
<div
  className="mock-provider">
  <button
    className="tc-action">
    My article
  </button>
</div>
`;
